File: src/orchestrator.py
"""
orchestrator.py

PDF → Markdown 変換パイプラインのエントリーポイント。
"""

import argparse
import logging
import os
import sys
from concurrent.futures import ThreadPoolExecutor, as_completed
from datetime import datetime

import yaml

from assembler import assemble_final_md
from gemini_client import get_gemini_patch
from markdown_drafter import draft_markdown
from page_store import store_page
from patch_applier import apply_patch
from rasterizer import rasterize_page
from text_extractor import extract_text_lines

# TODO: YAMLから読み込んだ値を各種関数に渡すようにする。現状は個別ファイル内でハードコードされている


def load_config(config_path):
    with open(config_path, "r", encoding="utf-8") as f:
        return yaml.safe_load(f)


def setup_logger(log_dir, log_level):
    os.makedirs(log_dir, exist_ok=True)
    timestamp = datetime.now().strftime("%Y%m%d_%H%M")
    log_file = os.path.join(log_dir, f"run_{timestamp}.log")
    log_fmt = "[%(asctime)s] %(levelname)s %(name)s - %(message)s"
    logging.basicConfig(
        level=getattr(logging, log_level.upper(), "INFO"),
        format=log_fmt,
        handlers=[
            logging.StreamHandler(sys.stdout),
            logging.FileHandler(log_file, encoding="utf-8"),
        ],
    )


def parse_args():
    parser = argparse.ArgumentParser(description="PDF → Markdown 変換パイプライン")
    parser.add_argument("pdf_path", type=str, help="入力PDFファイルパス")
    parser.add_argument(
        "--config", type=str, default="configs/config.yaml", help="設定ファイルパス"
    )
    parser.add_argument("--workers", type=int, help="並列ワーカー数")
    parser.add_argument("--dpi", type=int, help="画像DPI")
    parser.add_argument(
        "--resume-from", type=int, default=1, help="このページ番号から再開（1始まり）"
    )
    return parser.parse_args()


def main():
    args = parse_args()
    config = load_config(args.config)

    # CLI引数で上書き
    workers = args.workers if args.workers else config.get("workers", 4)
    dpi = args.dpi if args.dpi else config.get("dpi", 300)
    log_level = config.get("log_level", "INFO")
    output_dir = config.get("output_dir", "output")

    # タイムスタンプ付き出力ディレクトリ
    timestamp = datetime.now().strftime("%Y%m%d_%H%M")
    run_output_dir = os.path.join(output_dir, timestamp)
    pages_dir = os.path.join(run_output_dir, "pages")
    images_dir = os.path.join(run_output_dir, "images")
    os.makedirs(pages_dir, exist_ok=True)
    os.makedirs(images_dir, exist_ok=True)

    # ログ初期化
    setup_logger("logs", log_level)
    logger = logging.getLogger("orchestrator")
    logger.info(f"Started processing: {args.pdf_path}")

    # PDFページ数取得
    import pdfplumber

    with pdfplumber.open(args.pdf_path) as pdf:
        num_pages = len(pdf.pages)

    logger.info(f"Total pages: {num_pages}")

    # 再開ページ番号（1始まり→0始まりに変換）
    resume_from = max(args.resume_from - 1, 0)

    # Page Worker 処理
    gemini_model = config.get("gemini_model", "gemini-2.5-flash-preview-04-17")
    max_retries = config.get("max_retries", 3)
    retry_backoff = config.get("retry_backoff", 2)
    def process_page(page_num):
        logger.info(f"Started page {page_num + 1}")
        # 1. テキスト抽出
        draft_md = extract_text_lines(args.pdf_path, page_num, dpi)
        # 2. Markdown整形
        md = draft_markdown(draft_md)
        # 3. 画像生成
        image_path = rasterize_page(args.pdf_path, page_num, dpi, images_dir)
        # 4. Geminiパッチ取得
        answer, price = get_gemini_patch(md, image_path, gemini_model, max_retries, retry_backoff)
        logger.info(f"Gemini patch cost: ${price:.4f}")
        # 5. パッチ適用
        fixed_md = apply_patch(md, answer)
        # 6. ページ保存
        page_md_path = store_page(fixed_md, page_num, pages_dir, image_path)
        logger.info(f"Finished page {page_num + 1}")
        return page_md_path, price

    # 並列実行
    page_md_paths = []
    total_price = 0.0
    with ThreadPoolExecutor(max_workers=workers) as executor:
        # resume_from未満のページはスキップ
        futures = {
            executor.submit(process_page, i): i for i in range(resume_from, num_pages)
        }
        for future in as_completed(futures):
            page_md_path, price = future.result()
            page_md_paths.append(page_md_path)
            total_price += price
    logger.info(f"Total cost: ${total_price:.4f}")

    # ページ順に並べ替え
    page_md_paths.sort()

    # アセンブル
    final_md_path = assemble_final_md(page_md_paths, run_output_dir, images_dir)
    logger.info(f"Output: {final_md_path}")


if __name__ == "__main__":
    main()

File: src/orchestrator.ts
import { mkdirSync } from "node:fs";
import path from "node:path";
import pLimit from "p-limit";
import { assembleFinalMd } from "./assembler.js";
import { parseConfig } from "./config.js";
import { getGeminiPatch } from "./gemini-client.js";
import { initLogger, logger } from "./logger.js";
import { draftMarkdown } from "./markdown-drafter.js";
import { type MermaidFailure, validateAndFixMermaid } from "./mermaid-validator.js";
import { storePage } from "./page-store.js";
import { applyPatch } from "./patch-applier.js";
import { rasterizePage } from "./rasterizer.js";
import { extractTextLines, getPageCount } from "./text-extractor.js";

function formatTimestamp(): string {
	const now = new Date();
	const y = now.getFullYear();
	const mo = String(now.getMonth() + 1).padStart(2, "0");
	const d = String(now.getDate()).padStart(2, "0");
	const h = String(now.getHours()).padStart(2, "0");
	const mi = String(now.getMinutes()).padStart(2, "0");
	return `${y}${mo}${d}_${h}${mi}`;
}

function renderProgress(done: number, total: number, cost: number): void {
	const costStr = `$${cost.toFixed(4)}`;
	if (!process.stdout.isTTY) {
		if (done === total) {
			console.log(`Pages ${done}/${total}  cost: ${costStr}`);
		}
		return;
	}
	const width = 24;
	const filled = Math.min(Math.round((done / total) * width), width);
	const bar = "=".repeat(filled) + " ".repeat(width - filled);
	process.stdout.write(`\rPages [${bar}] ${done}/${total}  cost: ${costStr}`);
	if (done === total) {
		process.stdout.write("\n");
	}
}

async function main() {
	const config = parseConfig();
	const timestamp = formatTimestamp();

	// 中間生成物は /tmp/pdf-reader/<timestamp>/ に置く
	const runOutputDir = path.join("/tmp", "pdf-reader", timestamp);
	const pagesDir = path.join(runOutputDir, "pages");
	const imagesDir = path.join(runOutputDir, "images");
	mkdirSync(pagesDir, { recursive: true });
	mkdirSync(imagesDir, { recursive: true });

	// ログ初期化(中間生成物と同じdirへ)
	initLogger(runOutputDir, config.logLevel);
	logger.info(`Started processing: ${config.pdfPath}`);
	logger.info(`Intermediate dir: ${runOutputDir}`);

	// PDFページ数取得
	const numPages = await getPageCount(config.pdfPath);
	logger.info(`Total pages: ${numPages}`);

	// 再開ページ番号(1始まり→0始まりに変換)
	const resumeFrom = Math.max(config.resumeFrom - 1, 0);
	const totalToProcess = numPages - resumeFrom;
	let completed = 0;
	let runningCost = 0;
	const mermaidFailures: MermaidFailure[] = [];

	renderProgress(0, totalToProcess, 0);

	// Page Worker 処理
	async function processPage(pageNum: number): Promise<[string, number]> {
		logger.file("debug", `Started page ${pageNum + 1}`);

		// 1. テキスト抽出
		const draftMd = await extractTextLines(config.pdfPath, pageNum);
		// 2. Markdown整形
		const md = draftMarkdown(draftMd);
		// 3. 画像生成
		const imagePath = await rasterizePage(config.pdfPath, pageNum, config.dpi, imagesDir);
		// 4. Geminiパッチ取得
		const [answer, patchCost] = await getGeminiPatch(
			md,
			imagePath,
			config.geminiModel,
			config.maxRetries,
			config.retryBackoff,
		);
		logger.file("debug", `Gemini patch cost: $${patchCost.toFixed(4)}`);
		// 5. パッチ適用
		const fixedMd = applyPatch(md, answer);
		// 5.5 mermaid検証・修正
		const {
			md: validatedMd,
			cost: mermaidCost,
			failures,
		} = await validateAndFixMermaid(
			fixedMd,
			config.geminiModel,
			config.mermaidMaxRetries,
			config.retryBackoff,
			pageNum,
		);
		if (failures.length > 0) {
			mermaidFailures.push(...failures);
		}
		const price = patchCost + mermaidCost;
		// 6. ページ保存
		const pageMdPath = storePage(validatedMd, pageNum, pagesDir);
		completed++;
		runningCost += price;
		logger.file("info", `Finished page ${pageNum + 1} (${completed}/${totalToProcess})`);
		renderProgress(completed, totalToProcess, runningCost);

		return [pageMdPath, price];
	}

	// 並列実行
	const limit = pLimit(config.workers);
	const pages = Array.from({ length: numPages - resumeFrom }, (_, i) => i + resumeFrom);
	const results = await Promise.all(pages.map((pageNum) => limit(() => processPage(pageNum))));

	let totalPrice = 0;
	const pageMdPaths: string[] = [];
	for (const [pageMdPath, price] of results) {
		pageMdPaths.push(pageMdPath);
		totalPrice += price;
	}

	logger.info(`Total cost: $${totalPrice.toFixed(4)}`);

	// ページ順に並べ替え
	pageMdPaths.sort();

	// アセンブル(入力PDFと同じdirへ書き出す)
	const { purePath, withImagesPath } = assembleFinalMd({
		pageMdPaths,
		imagesDir,
		pdfPath: config.pdfPath,
		withImages: config.withImages,
	});
	logger.info(`Output: ${purePath}`);
	if (withImagesPath) {
		logger.info(`Output (with images): ${withImagesPath}`);
	}

	if (mermaidFailures.length > 0) {
		logger.warn(`Mermaid修正失敗: ${mermaidFailures.length}件`);
		for (const f of mermaidFailures) {
			const firstErrLine = f.error.split("\n").find((l) => l.trim().length > 0) ?? "";
			logger.warn(`  p.${f.pageNum + 1}: ${firstErrLine}`);
		}
	}
}

main().catch((err) => {
	logger.error(`Fatal: ${err instanceof Error ? (err.stack ?? err.message) : String(err)}`);
	process.exit(1);
});
